csharp-mode.el

```diff
@@ -2077,1076 +1837,282 @@ to the beginning of the prior namespace.
+                        "\\[" optional-space
+                        ;; type
+                        "\\([^\]]+\\)" optional-space
+                        identifier
+                        ;; closing brackets
+                        "\\]"
+                        "\\)"
+                        optional-space "{" optional-space
+                        ;; unless we are super-specific and expect the accesors,
+                        ;; lots of weird things gets slurped into the name.
+                        ;; including the accessors themselves.
+                        (regexp-opt '("get" "set"))) 1)
+          (list "event"
+                (concat bol
+                        access-modifier-list "+"
+                        optional-space "event" optional-space
+                        "\\("
+                        return-type space
+                        generic-identifier
+                        "\\)"
+                        optional-space
+                        ";") 1))))
+
+(defun csharp--imenu-get-pos (pair)
+  (let ((pos (cdr pair)))
+    (if (markerp pos)
+        (marker-position pos)
+      pos)))
+
+(defun csharp--imenu-get-container (item containers previous)
+  (if (not containers)
+      previous
+    (let* ((item-pos (csharp--imenu-get-pos item))
+           (container (car containers))
+           (container-pos (csharp--imenu-get-pos container))
+           (rest      (cdr containers)))
+      (if (and container-pos
+               (< item-pos container-pos))
+          previous
+        (csharp--imenu-get-container item rest container)))))
+
+(defun csharp--imenu-get-container-name (item containers)
+  (let ((container (csharp--imenu-get-container item containers nil)))
+    (if (not container)
+        nil
+      (let* (;; namespace
+             (container-p1 (car (split-string (car container))))
+             ;; class/interface
+             (container-p2 (cadr (split-string (car container)))))
+        ;; use p1 (namespace) when there is no p2
+        (if container-p2
+            container-p2
+          container-p1)))))
+
+(defun csharp--imenu-sort (items)
+  (sort items (lambda (item1 item2)
+                (string< (car item1) (car item2)))))
+
+(defun csharp--imenu-get-class-name (class namespaces)
+  (let ((namespace (csharp--imenu-get-container-name class namespaces))
+        (class-name (car class)))
+    (if (not namespace)
+        class-name
+      ;; reformat to include namespace
+      (let* ((words (split-string class-name))
+             (type  (car words))
+             (name  (cadr words)))
+        (concat type " " namespace "." name)))))
+
+(defun csharp--imenu-get-class-nodes (classes namespaces)
+  "Creates a new alist with classes as root nodes with namespaces added."
+
+  (mapcar (lambda (class)
+            (let* ((class-name (csharp--imenu-get-class-name class namespaces))
+                   (class-pos  (cdr class)))
+              (cons class-name
+                    (list
```

Comment:
Oh right. But while that is indeed technically correct, what we're trying to construct here is a alist-entry (a cons pair) where the value-portion itself is a list of such alist-entries.

As such, I think the current implementation in a better and more direct way reflects that, rather than incidentally ending up with the same construct using less obvious functions.

This might be good stuff to clear up in a comment though, and I'll be sure to do that.